python_modules/dagster/dagster/_components/__init__.py

```diff
@@ -26,7 +40,29 @@ def generate_files(cls) -> None:
         raise NotImplementedError()
 
     @abstractmethod
-    def build_defs(self) -> "Definitions": ...
+    def build_defs(self, context: "ComponentLoadContext") -> "Definitions": ...
+
+
+class AssetsComponent(Component):
+    path: Path
+    specs: Sequence[AssetSpec]
+
+    def __init__(self, path: Union[str, Path], specs: Optional[Sequence[AssetSpec]] = None):
+        self.path = Path(path)
+        self.specs = specs or [AssetSpec(key=self.path.stem)]
+
+
+class ComponentCollection(Component):
+    def __init__(self, component_type: Type[Component], components: Sequence[Component]):
+        self.component_type = component_type
+        self.components = check.list_param(components, "components", of_type=component_type)
```

Comment:
I avoided doing that to start because there are some non-trivial init things we're doing (for example in AssetsComponent, we construct a default value for self.specs based on self.path). That sort of thing is possible to do with dataclasses but not really the "happy path" there